_published/0017-azuredevops-repos-and-not-projects.md

```diff
@@ -0,0 +1,259 @@
+(undocumented) defaults that use the same variables behind the scenes, will be
+fine. Perhaps some Docker images are pushed to a new address, but as long as
+that new registry address is writable and wherever that address is referenced
+uses those above built-in variables then will work fine.
+
+The issue is where users have hardcoded these URLs into their continuous
+deployments. The Docker images will be pushed to a new location, but the old
+location is still what's referenced, causing their builds to break.
+
+This information needs to go out, preferably before this is released, to all
+known users of wharf-provider-azuredevops telling them to start using
+`${REPO_GROUP}` and `${REPO_NAME}` instead.
+
+For the unknown users, we should try and word the CHANGELOG.md as clear as
+possible declaring these impacts. We should also write a document on
+<https://iver-wharf.github.io/> about upgrading wharf-provider-azuredevops from
+v1 to v2, where we explain these backward incompatibilities as well.
+
+## Alternative solutions
+
+For the sake of resolving
+<https://github.com/iver-wharf/wharf-provider-azuredevops/issues/24> we could
+settle with still importing 1 Wharf project per Azure DevOps project, but
+this is so tightly coupled so I [@jilleJr] think it's best to tackle this
+inconsistency here and now. This RFC is not meant to fix that bug, but instead
+designed so that bug is fixed as a consequence.
+
+## Future possibilities
+
+- We can update the README.md inside wharf-provider-azuredevops :)
+
+  ```diff
+  -Import Wharf projects from Azure DevOps repositories. Mainly focused on
+  -importing from self hosted Azure DevOps instances, importing from
+  -dev.azure.com is not well tested.
+  +Import Wharf projects from Azure DevOps repositories. Tested on importing
+  +both from self hosted Azure DevOps instances as well as from <dev.azure.com>.
+  ```
+
+- For this issue: [Map projects per ID: update importers code #6](https://github.com/iver-wharf/wharf-provider-azuredevops/issues/6)
+  we can import based on the Azure DevOps repository's GUIDs instead of the
+  project GUIDs, which would be more accurate and would mean less backward
+  compatibility migrations for the future, as this "importing repos and not
+  projects" change seems inevitable anyway.
+
+## Unresolved questions
+
+- How to handle the naming?
+
+  There are three main naming styles:
+
+  | ID  | Project group                 | Project name     |
+  | --- | -------------                 | ------------     |
+  | A   | `{azure-org}/{azure-project}` |`{azure-repo}`    |
+  | B   | `{azure-org}`                 |`{azure-repo}`    |
+  | C   | `{azure-org}`                 |`{azure-project}` |
+
+  The latter, naming C, is used in wharf-provider-azuredevops v1. For v2, how
+  do we handle the naming when there are more than 1 code repository?
+
+  1. Always use the naming A.
```

Comment:
I am voting for first option. Adding conditions to naming as it is proposed in 2nd and 3rd points can make it less readable and generate more bugs.

Reply:
You're right. My only intention with this "conditional naming" was so the previously imported Wharf projects keep their names, and to try and avoid double naming. But if it's consistently double-naming then at least it's easier to understand for end-users.

I'm fond of this argument, so I'll update the RFC to use option 1.

Reply:
Updated